index.js

```diff
@@ -10,30 +11,36 @@ const getHtmlFiles = async (directory) => {
   return files.map((file) => file.fullPath)
 }
 
+const CRITICAL_CONCURRENCY_LIMIT = 4
+
 module.exports = {
   onPostBuild: async ({ inputs, constants, utils }) => {
     const htmlFiles = await getHtmlFiles(constants.PUBLISH_DIR)
+    const criticalOptions = htmlFiles.map((filePath) => ({
+      base: constants.PUBLISH_DIR,
+      // Overwrite files by passing the same path for `src` and `target`.
+      src: filePath,
+      target: filePath,
+      inline: true,
+      minify: inputs.minify,
+      extract: inputs.extract,
+      dimensions: inputs.dimensions
+    }))
 
     try {
       // Ignore penthouse/puppeteer max listener warnings.
       // See https://github.com/pocketjoso/penthouse/issues/250.
       // One penthouse call is made per page and per screen resolution.
-      process.setMaxListeners(htmlFiles.length * inputs.dimensions.length + 1)
+      process.setMaxListeners(
+        CRITICAL_CONCURRENCY_LIMIT * inputs.dimensions.length + 1
+      )
 
-      const inlineCriticalPromises = htmlFiles.map((filePath) =>
-        critical.generate({
-          base: constants.PUBLISH_DIR,
-          // Overwrite files by passing the same path for `src` and `target`.
-          src: filePath,
-          target: filePath,
-          inline: true,
-          minify: inputs.minify,
-          extract: inputs.extract,
-          dimensions: inputs.dimensions
-        })
+      await asyncPool(
```

Comment:
The “algorithm” is super sweet: https://github.com/rxaviers/async-pool/blob/master/lib/es7.js